lua/lsp/null-ls/init.lua

```diff
@@ -0,0 +1,78 @@
+local M = {}
+
+local providers_by_ft = {}
+
+local function list_provider_names(providers, options)
+  options = options or {}
+  local names = {}
+
+  local filter = options.filter or "supported"
+  for name, _ in pairs(providers[filter]) do
+    table.insert(names, name)
+  end
+
+  return names
+end
+
+function M.list_provider_names(filetype, options)
+  local names = {}
+
+  if not providers_by_ft[filetype] then
+    return names
+  end
+
+  for _, providers in pairs(providers_by_ft[filetype]) do
+    vim.list_extend(names, list_provider_names(providers, options))
+  end
+
+  return names
+end
+
+function M.list_supported_formatter_names(filetype)
+  if not providers_by_ft[filetype] then
+    return {}
+  end
+  return list_provider_names(providers_by_ft[filetype].formatters, { filter = "supported" })
+end
+
+function M.list_supported_linter_names(filetype)
+  if not providers_by_ft[filetype] then
+    return {}
+  end
+  return list_provider_names(providers_by_ft[filetype].linters, { filter = "supported" })
+end
+
+function M.list_unsupported_formatter_names(filetype)
+  if not providers_by_ft[filetype] then
+    return {}
+  end
+  return list_provider_names(providers_by_ft[filetype].formatters, { filter = "unsupported" })
+end
+
+function M.list_unsupported_linter_names(filetype)
+  if not providers_by_ft[filetype] then
+    return {}
+  end
+  return list_provider_names(providers_by_ft[filetype].linters, { filter = "unsupported" })
+end
```

Comment:
nit: move these to their respective module and also provide an installer function to replace this 
```lua
  providers_by_ft[filetype] = {}
  providers_by_ft[filetype].formatters = null_formatters.list_configured(lvim.lang[filetype].formatters)
  providers_by_ft[filetype].linters = null_linters.list_configured(lvim.lang[filetype].linters)
  null_ls.register { sources = providers_by_ft[filetype].formatters.supported }
  null_ls.register { sources = providers_by_ft[filetype].linters.supported }
```
It will allow for changing the implementation details, if necessary, without modifying the interface.